api/src/main/java/org/apache/iceberg/DataFile.java

```diff
@@ -45,7 +45,8 @@ static StructType getType(StructType partitionType) {
         required(102, "partition", partitionType),
         required(103, "record_count", LongType.get()),
         required(104, "file_size_in_bytes", LongType.get()),
-        required(105, "block_size_in_bytes", LongType.get()),
+        // Note that although this field is optional, we need to always write it for older readers
+        optional(105, "block_size_in_bytes", LongType.get()),
```

Comment:
This is my fault, but I don't think it is safe to make this optional. It is going to change the encoding and older readers may fail. I think we should keep it as a require field and make the other changes in this file.